### ISSUE TITLE
Is there / Should there be a way to build widget inside async function?

### ISSUE THREAD

**Issue by kergefarkas** [flutter/flutter#16062]
## Steps to Reproduce

I have an async function that scrapes data from a website. Once the parsing is done, I want to build a widget using that information, but even though the data is present (I can see it in the debug window), the will not be built and I get the following error:

> I/flutter ( 9319): ══╡ EXCEPTION CAUGHT BY RENDERING LIBRARY ╞═════════════════════════════════════════════════════════
> I/flutter ( 9319): The following assertion was thrown during performLayout():
> I/flutter ( 9319): type '_Future' is not a subtype of type 'Widget' of 'child' where
> I/flutter ( 9319):   _Future is from dart:async
> I/flutter ( 9319):   Widget is from package:flutter/src/widgets/framework.dart
> 
> I/flutter ( 9567): When the exception was thrown, this was the stack:
> I/flutter ( 9567): #0      SliverChildBuilderDelegate.build (package:flutter/src/widgets/sliver.dart:197)
> I/flutter ( 9567): #1      SliverMultiBoxAdaptorElement._build.<anonymous closure> (package:flutter/src/widgets/sliver.dart:716)
> I/flutter ( 9567): #2      _HashMap.putIfAbsent (dart:collection-patch/dart:collection/collection_patch.dart:143)
> I/flutter ( 9567): #3      SliverMultiBoxAdaptorElement._build (package:flutter/src/widgets/sliver.dart:716)
> I/flutter ( 9567): #4      SliverMultiBoxAdaptorElement.createChild.<anonymous closure> (package:flutter/src/widgets/sliver.dart:729)
> I/flutter ( 9567): #5      BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2205)
> I/flutter ( 9567): #6      SliverMultiBoxAdaptorElement.createChild (package:flutter/src/widgets/sliver.dart:722)
> I/flutter ( 9567): #7      RenderSliverMultiBoxAdaptor._createOrObtainChild.<anonymous closure> (package:flutter/src/rendering/sliver_multi_box_adaptor.dart:254)
> I/flutter ( 9567): #8      RenderObject.invokeLayoutCallback.<anonymous closure> (package:flutter/src/rendering/object.dart:1666)
> I/flutter ( 9567): #9      PipelineOwner._enableMutationsToDirtySubtrees (package:flutter/src/rendering/object.dart:740)
> I/flutter ( 9567): #10     RenderObject.invokeLayoutCallback (package:flutter/src/rendering/object.dart:1666)
> I/flutter ( 9567): #11     RenderSliverMultiBoxAdaptor._createOrObtainChild (package:flutter/src/rendering/sliver_multi_box_adaptor.dart:243)
> I/flutter ( 9567): #12     RenderSliverMultiBoxAdaptor.addInitialChild (package:flutter/src/rendering/sliver_multi_box_adaptor.dart:351)
> I/flutter ( 9567): #13     RenderSliverList.performLayout (package:flutter/src/rendering/sliver_list.dart:78)
> I/flutter ( 9567): #14     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #15     RenderSliverPadding.performLayout (package:flutter/src/rendering/sliver_padding.dart:182)
> I/flutter ( 9567): #16     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #17     RenderViewportBase.layoutChildSequence (package:flutter/src/rendering/viewport.dart:285)
> I/flutter ( 9567): #18     RenderViewport._attemptLayout (package:flutter/src/rendering/viewport.dart:1008)
> I/flutter ( 9567): #19     RenderViewport.performLayout (package:flutter/src/rendering/viewport.dart:932)
> I/flutter ( 9567): #20     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #21     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107)
> I/flutter ( 9567): #22     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #23     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107)
> I/flutter ( 9567): #24     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #25     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107)
> I/flutter ( 9567): #26     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #27     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107)
> I/flutter ( 9567): #28     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #29     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107)
> I/flutter ( 9567): #30     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #31     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107)
> I/flutter ( 9567): #32     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #33     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107)
> I/flutter ( 9567): #34     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #35     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107)
> I/flutter ( 9567): #36     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #37     MultiChildLayoutDelegate.layoutChild (package:flutter/src/rendering/custom_layout.dart:124)
> I/flutter ( 9567): #38     _ScaffoldLayout.performLayout (package:flutter/src/material/scaffold.dart:257)
> I/flutter ( 9567): #39     MultiChildLayoutDelegate._callPerformLayout (package:flutter/src/rendering/custom_layout.dart:194)
> I/flutter ( 9567): #40     RenderCustomMultiChildLayoutBox.performLayout (package:flutter/src/rendering/custom_layout.dart:338)
> I/flutter ( 9567): #41     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #42     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107)
> I/flutter ( 9567): #43     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #44     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107)
> I/flutter ( 9567): #45     _RenderCustomClip.performLayout (package:flutter/src/rendering/proxy_box.dart:1141)
> I/flutter ( 9567): #46     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #47     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107)
> I/flutter ( 9567): #48     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #49     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107)
> I/flutter ( 9567): #50     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #51     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107)
> I/flutter ( 9567): #52     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #53     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107)
> I/flutter ( 9567): #54     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #55     RenderOffstage.performLayout (package:flutter/src/rendering/proxy_box.dart:2711)
> I/flutter ( 9567): #56     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #57     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:107)
> I/flutter ( 9567): #58     RenderObject.layout (package:flutter/src/rendering/object.dart:1570)
> I/flutter ( 9567): #59     RenderStack.performLayout (package:flutter/src/rendering/stack.dart:509)
> I/flutter ( 9567): #60     RenderObject._layoutWithoutResize (package:flutter/src/rendering/object.dart:1445)
> I/flutter ( 9567): #61     PipelineOwner.flushLayout (package:flutter/src/rendering/object.dart:709)
> I/flutter ( 9567): #62     BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:270)
> I/flutter ( 9567): #63     BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:622)
> I/flutter ( 9567): #64     BindingBase&GestureBinding&ServicesBinding&SchedulerBinding&PaintingBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:208)
> I/flutter ( 9567): #65     BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:990)
> I/flutter ( 9567): #66     BindingBase&GestureBinding&ServicesBinding&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:930)
> I/flutter ( 9567): #67     BindingBase&GestureBinding&ServicesBinding&SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:842)
> I/flutter ( 9567): #68     _invoke (file:///b/build/slave/Linux_Engine/build/src/flutter/lib/ui/hooks.dart:120)
> I/flutter ( 9567): #69     _drawFrame (file:///b/build/slave/Linux_Engine/build/src/flutter/lib/ui/hooks.dart:109)
> 
> I/flutter ( 9567): The following RenderObject was being processed when the exception was fired:
> I/flutter ( 9567):   RenderSliverList#042ea relayoutBoundary=up2 NEEDS-LAYOUT NEEDS-PAINT
> I/flutter ( 9567):   creator: SliverList ← MediaQuery ← SliverPadding ← Viewport ← _ScrollableScope ←
> I/flutter ( 9567):   IgnorePointer-[GlobalKey#4ab34] ← Semantics ← Listener ← _GestureSemantics ←
> I/flutter ( 9567):   RawGestureDetector-[LabeledGlobalKey<RawGestureDetectorState>#be83c] ←
> I/flutter ( 9567):   _ExcludableScrollSemantics-[GlobalKey#29c09] ← RepaintBoundary ← ⋯
> I/flutter ( 9567):   parentData: paintOffset=Offset(0.0, 0.0) (can use size)
> I/flutter ( 9567):   constraints: SliverConstraints(AxisDirection.down, GrowthDirection.forward, ScrollDirection.idle,
> I/flutter ( 9567):   scrollOffset: 0.0, remainingPaintExtent: 694.9, crossAxisExtent: 411.4, crossAxisDirection:
> I/flutter ( 9567):   AxisDirection.right, viewportMainAxisExtent: 694.9)
> I/flutter ( 9567):   geometry: null
> I/flutter ( 9567):   no children current live
> I/flutter ( 9567): This RenderObject has no descendants.

Here is my code:

```
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:html/parser.dart' show parse;
import 'package:html/dom.dart' as dom;
import 'utilities.dart';

fetchBasicData(String URL) async {
  final response = await http.get(URL);
  var document = parse(response.body);

  var result = new List<dom.Node>();
  result = document.getElementsByClassName('datas-nev');
  var dogName = result[0].nodes[1].toString();
  result = document.getElementsByClassName('datas-tipus');
  var dogBreed = result[0].nodes[1].toString();
  result = document.getElementsByClassName('datas-nem');
  var dogGender = result[0].nodes[1].toString();
  result = document.getElementsByClassName('datas-szin');
  var dogColor = result[0].nodes[1].toString();
  result = document.getElementsByClassName('datas-kor');
  var dogAge = result[0].nodes[1].toString();

  result = document.getElementsByClassName('pirobox_gall');
  String imageLink;
  imageLink = urlPrefix + result[0].nodes[0].attributes.values.first;

  Dog myDog = new Dog.basic(
      URL,
      dogName,
      dogBreed,
      dogGender,
      dogColor,
      dogAge,
      imageLink);

  return new SizedBox(
      width: 500.0,
      height: 400.0,
      child: new Card(
        child: new Column(
          mainAxisAlignment: MainAxisAlignment.start,
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            //Header image row
            new Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                new Expanded(
                    child: new FutureBuilder(
                        future: fetchPortrait(myDog.imageLink),
                        builder: (context, snapshot) {
                          if (snapshot.hasData) {
                            return new Image.network(
                              snapshot.data,
                              fit: BoxFit.scaleDown,
                            );
                          } else {
                            if (snapshot.hasError) {
                              return new Text('Hiba');
                            }
                          }
                          return new Center(
                            child: new CircularProgressIndicator(),
                          );
                        }))
              ],
            ), //Header image row
            new Row(
              children: <Widget>[
                new Expanded(
                    child: new Text(
                      myDog.dogName,
                      style: new TextStyle(
                          fontWeight: FontWeight.bold,
                          fontSize: 18.0,
                          color: Colors.black),
                    ))
              ],
            ),
            new Row(
              children: <Widget>[
                new Expanded(
                    child: new Text(myDog.dogColor +
                        'színű, ' +
                        myDog.dogBreed +
                        ' ' +
                        myDog.dogGender))
              ],
            ),
            new Row(
              children: <Widget>[
                new Expanded(child: new Text('Kora: kb. ' + myDog.dogAge))
              ],
            )
          ],
        ),
      ),
    );
}
```

Circumvention of the problem is possible, by putting the data fetching in FutureBuilder widgets, for example the way the fetchPortrait function is used here.

But I think the widget building inside the async function should be allowed, as proper data preparation (like using try-catch to initialize some default value in case some of the data is not available) could take care of possible problems.

[flutterv.txt](https://github.com/flutter/flutter/files/1859932/flutterv.txt)

**Comment by mehmetf** [comment]
What does the UI do while the data is being fetched? I assume you are calling this from build() of another Widget.

The correct paradigm to use for this is setState().
1. If data is not loaded, return a loading screen while firing off your async data getter.
2. When future is complete, setState() with the new data.
3. Rebuild will now see the data being loaded and show you the right screen.

**Comment by patelkrunal78** [comment]
The best way to do this is to use a FutureBuilder.

From the FutureBuilder documentation:

new FutureBuilder<String>(
  future: _calculation, // a Future<String> or null
  builder: (BuildContext context, AsyncSnapshot<String> snapshot) {
    switch (snapshot.connectionState) {
      case ConnectionState.none: return new Text('Press button to start');
      case ConnectionState.waiting: return new Text('Awaiting result...');
      default:
        if (snapshot.hasError)
          return new Text('Error: ${snapshot.error}');
        else
          return new Text('Result: ${snapshot.data}');
    }
  },
)
The other thing is that you're building your widget outside of the State.build method and saving the widget itself, which is an anti-pattern. You should be actually building the widgets each time in the build method.

You could get this to work without FutureBuilder, but you should save the result of the http call (processed appropriately) and then use the data within your build function.

See this, but note that using a FutureBuilder is a better way to do this and I'm just providing this for you to learn.

import 'dart:async';
import 'dart:convert';

import 'package:flutter/material.dart';

void main() {
  runApp(new MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter demo',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'async demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List data;

  @override
  initState() {
    super.initState();

    new Future<String>.delayed(new Duration(seconds: 5), () => '["123", "456", "789"]').then((String value) {
      setState(() {
        data = json.decode(value);
      });
    });
  }

  @override
  Widget build(BuildContext context) {
    if (data == null) {
      return new Scaffold(
        appBar: new AppBar(
          title: new Text("Loading..."),
        ),
      );
    } else {
      return new Scaffold(
        appBar: new AppBar(
          title: new Text(widget.title),
        ),
        body: new Center(
          child: new ListView(
            children: data
                .map((data) => new ListTile(
                      title: new Text("one element"),
                      subtitle: new Text(data),
                    ))
                .toList(),
          ),
        ),
      );
    }
  }
}


Find Code At [Here](https://icetutor.com/question/flutter-render-widget-after-async-call)
